### PR TITLE
DCOS-10979: Networking section in Service Create modal

### PR DESCRIPTION
**Pull request by jfurrow**
This introduces the start of the Networking section. So sorry for the huge PR! A lot of it is tests, as I wanted to make sure that we were catching all the corner cases I was discovering.

It adds JSONParser, JSONReducer and a FormReducer for portDefinitions and network type. And only JSONParser and JSONReducer for portMappings (since the form only handles one entry).

Note: That the portMappings take precedence in the JSON Editor. This is by design!

### CODE REVIEW

**Comment by mlunoe** [comment]
It seems there is something wrong with the path for `serviceForm/PortDefinitions`.
It's located under `ServiceForm/` (capitol S), not `serviceForm/` (lower cased s):


And this needs a rebase

**Comment by jfurrow** [comment]
@mlunoe Rebased and fixed the broken UI

**Comment by mlunoe** [comment]
**mesosphere-ci** Retest this please

**Comment by mlunoe** [comment]
**mesosphere-ci** Retest this please

**Comment by mlunoe** [comment]
I'm still seeing the same failure here @jfurrow. Maybe you have an idea why this is still failing.
Maybe it couldn't handle the subtle change of having a capital letter changed to a lower cased?

**Comment by orlandohohmeier** [comment]
@mlunoe @jfurrow is there anything we can do to help you guys land this one? @nLight might has bandwidth to support you.

**Comment by orlandohohmeier** [comment]
The form must not remove the service port nor overwrite the vip setting when editing any of the other options. Please use the following examples for testing:

**Service Port**
```
{ "id": "serviceport",
  "container": {
    "type": "DOCKER",
    "docker": {
      "image": "busybox",
      "network": "BRIDGE",
      "portMappings": [
        { "hostPort": 5514, "containerPort": 514, "servicePort": 10000 }
      ],
      "forcePullImage":true
    }
  },
  "instances": 1,
  "cpus": 1,
  "mem": 128
}
```

**VIP**
```
{
  "id": "/vip",
  "cpus": 0.001,
  "instances": 1,
  "mem": 128,
  "portDefinitions": [
    {
      "port": 0,
      "protocol": "tcp",
      "labels": {
        "VIP_0": "1.1.1.1:0"
      }
    }
  ]
}
```

**Comment by Poltergeist** [comment]
We need a selection of the `virtual networks` that are available and also the ipAddress field needs to be populated in that case.

```JSON
{
  "id": "/appname",
  "cpus": 1,
  "mem": 128,
  "instances": 1,
  "container": {
    "docker": {
      "image": "nginx",
      "portMappings": [
        {
          "containerPort": 8080,
          "protocol": "tcp",
          "name": "host",
          "servicePort": 8080,
          "labels": {
            "VIP_0": "/appname:8080"
          }
        }
      ],
      "network": "USER"
    }
  },
  "ipAddress": {
    "networkName": "dcos"
  }
}
```

**Comment by nLight** [comment]
I thought that `applyPatch` should preserve the fields we don't handle in reducers //cc @wavesoft

**Comment by wavesoft** [comment]
.. assuming they are not part of something that we `delete` on the `baseConfig`

**Comment by nLight** [comment]
We delete only `labels` and `env`

**Comment by jfurrow** [comment]
This is amazing! Nice work, @mlunoe!

In my first pass, I noticed the following:
1. Disabling "port mapping" after entering a host port doesn't reset the `hostPort` to 0. To reproduce: Select Virtual Network, then add a service endpoint, then enable "port mapping," then disable "assign automatically," then enter a port number, then disable "port mapping."


2. There is a period followed by a comma when more than one item is disabled in the network types:


3. Errors for service endpoint names appear in the JSON editor but not in the form:

**Comment by rcorral** [comment]
Went over this in detail with @mlunoe, it looks 👌 👌 👌 💯 💯 💯 
Few things to fix, but I think this is almost ready to 🚢 

Awesome work!!

**Comment by mlunoe** [comment]
Ok, all comments addressed!

**Comment by orlandohohmeier** [comment]
**mesosphere-ci** retest this please 🙏

**Comment by orlandohohmeier** [comment]
@mlunoe the error handling should be fixed by #1470

**Comment by orlandohohmeier** [comment]
@mlunoe please squash bbf16d2

**Comment by mlunoe** [comment]
Commit squashed!

File: plugins/services/src/js/reducers/serviceForm/PortMappings.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  ADD_ITEM,
+  SET
+} from '../../../../../../src/js/constants/TransactionTypes';
+import Transaction from '../../../../../../src/js/structs/Transaction';
+
+module.exports = {
+  JSONReducer: null,
+
+  /**
+   * This parser must take precedence over PortDefinition parser, as it assumes
```
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/PortMappings.js]: Parses should not have any dependencies on other parsers. Let's refactor this in the next iteration.

**Comment by orlandohohmeier** [comment]
@mlunoe I've rebased this branch and pushed a fix for the VIP label handling.

**Comment by orlandohohmeier** [comment]
Let's get this one merged, now that all comments are address. Great work everyone! 💯✨🚢

**Comment by rcorral** [comment]
@mlunoe noticed a few things when using the UI.

1. TCP/UDP checkboxes don't do anything when on bridge mode
2. The "ASSIGN AUTOMATICALLY" checkbox doesn't work on bridge mode either
3. Picking "Virtual Netowork: dcos" causes a validation error in JSON. Switching back to bridge or host causes that validation error to still be shown.

Almost there!

**Comment by mlunoe** [comment]
1. and 2. is fixed!

**Comment by rcorral** [comment]
retest this please

**Comment by rcorral** [comment]
🙌 awesome work @mlunoe!!
